core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -243,6 +243,22 @@ class KafkaApis(val requestChannel: RequestChannel,
         s"${updateMetadataRequest.brokerEpoch()} smaller than the current broker epoch ${controller.brokerEpoch}")
       sendResponseExemptThrottle(request, new UpdateMetadataResponse(Errors.STALE_BROKER_EPOCH))
     } else {
+      def shouldUpdateInternalTopicPartitionCount(topicName: String) = {
+        val partitionsOfTopic = updateMetadataRequest.partitionStates.keySet.asScala.filter(_.topic == topicName)
+        if (partitionsOfTopic.isEmpty)
+          false
+        else {
+          val maxPartitionCountInRequest = partitionsOfTopic.map(_.partition).max
```

Comment:
The variable name is a little confusing. What we are actually querying is the maximum partition, not the count.